compiler/src/main/kotlin/io/github/aplcornell/viaduct/backends/aby/ABYCircuitCodeGenerator.kt

```diff
@@ -0,0 +1,599 @@
+                            valueToBigInt(argument.value, argument.type.elementType.value),
+                            BIT_LENGTH,
+                        )
+                    } else {
+                        builder.addStatement(
+                            "val %L = %L",
+                            inputName,
+                            argument.type.shape.new(context) { indices ->
+                                val value = CodeBlock.builder()
+                                value.add("%L", argument.value)
+                                indices.forEach { value.add("[%L]", it) }
+                                CodeBlock.of(
+                                    "%L.putCONSGate(%L, %L)",
+                                    protocolToAbyPartyCircuit(protocol),
+                                    valueToBigInt(value.build(), argument.type.elementType.value),
+                                    BIT_LENGTH,
+                                )
+                            },
+                        )
+                    }
+                }
+
+                else -> throw UnsupportedCommunicationException(argument.protocol, protocol, argument.sourceLocation)
+            }
+            CodeBlock.of("%N", inputName)
+        }
+        return Pair(builder.build(), values)
+    }
+
+    override fun export(
+        protocol: Protocol,
+        arguments: List<Argument>,
+    ): Pair<CodeBlock, List<CodeBlock>> {
+        require(protocol is ABY)
+        require(context.host in protocol.hosts)
+        val builder = CodeBlock.builder()
+        val outShareNames = arguments.map { argument ->
+            val outShareName = context.newTemporary(argument.value.toString() + "_outShare")
+            val outRole = when (argument.protocol) {
+                Local(protocol.server) -> Role.SERVER
+                Local(protocol.client) -> Role.CLIENT
+                Replication(setOf(protocol.server, protocol.client)) -> Role.ALL
+                else -> throw IllegalStateException()
+            }
+            builder.addStatement(
+                "val %N = %L.putOUTGate(%L, %L)",
+                outShareName,
+                protocolToAbyPartyCircuit(protocol),
+                argument.value,
+                roleToCodeBlock(outRole),
+            )
+            CodeBlock.of("%N", outShareName)
+        }
+        builder.addStatement(
+            "%L.execCircuit()",
+            protocolToABYPartyMap[ABYPair(protocol.server, protocol.client)],
+        )
+        val cleartextTmps = arguments.zip(outShareNames).filter { (arg, _) -> (context.host in arg.protocol.hosts) }
+            .map { (arg, outShareName) ->
+                val cleartextTmp = context.newTemporary("cleartextTmp")
+                builder.addStatement(
+                    "val %N = %L.getClearValue32().%L",
+                    cleartextTmp,
+                    outShareName,
+                    if (arg.type.elementType.value is IntegerType) {
```

Comment:
Should be pulled out into a separate function next to `.toShare`. Maybe `.fromShare`.